gpumon/gpumon_interface.ml

```diff
@@ -12,60 +12,147 @@
  * GNU Lesser General Public License for more details.
  *)
 
+open Rpc
+open Idl
+
 let service_name = "gpumon"
 let queue_name = Xcp_service.common_prefix ^ service_name
 let xml_path = "/var/xapi/" ^ service_name
 
+(** Uninterpreted string associated with the operation *)
 type debug_info = string
+[@@deriving rpcty]
+
+(* Domain ID of VM *)
 type domid = int
+[@@deriving rpcty]
+
+(** Reason for incompatibility *)
+type incompatibility_reason =
+  | Host_driver
+  | Guest_driver
+  | GPU
+  | Other
+[@@deriving rpcty]
 
-type incompatibility_reason = Host_driver | Guest_driver | GPU | Other
-type compatibility = Compatible | Incompatible of incompatibility_reason list
+(** Boolean: compatible? *)
+type compatibility =
+  | Compatible
+  | Incompatible of incompatibility_reason list
+[@@deriving rpcty]
 
+(** PCI identifier of physical GPU *)
 type pgpu_address = string
+[@@deriving rpcty]
+
+(** Metadata of Nvidia physical GPU *)
 type nvidia_pgpu_metadata = string
+[@@deriving rpcty]
+
+(** Metadata of Nvidia virtual GPU *)
 type nvidia_vgpu_metadata = string
+[@@deriving rpcty]
+
+(** List of Nvidia virtual GPU metadata records *)
+type nvidia_vgpu_metadata_list = nvidia_vgpu_metadata list
+[@@deriving rpcty]
+
+
+(** Error wrapper *)
+type gpu_errors =
```

Comment:
One thing I'm uncertain about is whether we need an error for compatibility issues. I assume these are already handled, but is it worth including one just in case?

Reply:
The compatibility errors are only for the Nvidia module at the moment, we will need new ones for ATI or Intel if ever needed. How to deal with it I think is open for discussion:

It is probably not worth splitting:

```ocaml
type nvml_error = 
   | NvmlInterfaceNotAvailable
   (** Exception raised when gpumon is unable to load the nvml nvidia library *)
   | NvmlFailure of string
   (** Exception raised by the c bindings to the nvml nvidia library*)
[@@deriving rpcty]

(** Error wrapper *)
 type gpu_errors =
   | Nvml of nvml_error
   (** Error raised by the Nvml library bindings *)
   | Gpumon_failure
   (** Default exception raised upon daemon failure *)
 [@@default Gpumon_failure]
[@@deriving rpcty]
```